src/Compilers/CSharp/Portable/Symbols/TupleTypeSymbol.cs

```diff
@@ -71,33 +72,83 @@ internal static TupleTypeSymbol ConstructTupleTypeSymbol(TupleTypeSymbol origina
+        /// Returns the nested type at a certain depth.
+        ///
+        /// For depth=0, just return the tuple type as-is.
+        /// For depth=1, returns the nested tuple type at position 8.
+        /// </summary>
+        public static NamedTypeSymbol GetNestedTupleType(NamedTypeSymbol tupleType, int depth)
+        {
+            NamedTypeSymbol found = tupleType;
+            for (int i = 0; i < depth; i++)
             {
-                var wellKnownTypeMember = GetTupleTypeMember(numElements, i);
-                var underlyingField = (FieldSymbol)Binder.GetWellKnownTypeMember(binder.Compilation, wellKnownTypeMember, diagnostics, syntax: syntax);
-                var field = new TupleFieldSymbol(elementNames.IsDefault ? UnderlyingMemberName(i) : elementNames[i],
-                                           this,
-                                           elementTypes[i],
-                                           underlyingField?.AsMember(underlyingType));
-                fieldSymbols.Add(field);
+                found = (NamedTypeSymbol)found.TypeArguments[7];
             }
+            return found;
+        }
+
+        /// <summary>
+        /// Returns the number of nestings required to represent numElements as nested ValueTuples.
+        /// For example, for 8 elements, you need 2 ValueTuples and the remainder (ie the size of the last nested ValueTuple) is 1.
+        /// </summary>
+        internal static int NumberOfValueTuples(int numElements, out int remainder)
+        {
+            remainder = (numElements - 1) % 7 + 1;
+            return (numElements - 1) / 7 + 1;
+        }
 
-            fields = fieldSymbols.ToImmutableAndFree();
+        /// <summary>
+        /// Produces the underlying ValueTuple corresponding to this list of element types.
+        ///
+        /// For 2 elementTypes of type Int32, returns ValueTuple&lt;Int32, Int32>
+        /// For 9 elementTypes, returns ValueTuple&lt;Int32, ..., ValueTuple&lt;Int32, Int32>>, and ValueTuple&lt;Int32, Int32> (largest first)
+        /// </summary>
+        private static NamedTypeSymbol GetTupleUnderlyingType(ImmutableArray<TypeSymbol> elementTypes, CSharpSyntaxNode syntax, Binder binder, DiagnosticBag diagnostics)
+        {
+            int numElements = elementTypes.Length;
+            int remainder;
+            int chainLength = NumberOfValueTuples(numElements, out remainder);
 
-            return underlyingType;
+            NamedTypeSymbol currentSymbol = default(NamedTypeSymbol);
+            NamedTypeSymbol firstTupleType = binder.GetWellKnownType(GetTupleType(remainder), diagnostics, syntax);
+            currentSymbol = firstTupleType.Construct(ImmutableArray.Create(elementTypes, (chainLength - 1) * 7, remainder));
+
+            for (int linkIndex = chainLength - 2; linkIndex >= 0; linkIndex--)
+            {
+                NamedTypeSymbol chainedTupleType = binder.GetWellKnownType(GetTupleType(8), diagnostics, syntax);
```

Comment:
This can be moved out of the loop, assuming we add an 'if' to execute the assignment and the loop only when there  is a nesting.